# Feature Scaling

In [1]:
import pandas as pd 
import numpy as np
import matplotlib.pyplot as plt
import seaborn as sns
     

In [2]:
pip install -U scikit-learn

  Using cached threadpoolctl-3.6.0-py3-none-any.whl.metadata (13 kB)
   ---------------------------------------- 0.0/8.9 MB ? eta -:--:--
   - -------------------------------------- 0.3/8.9 MB ? eta -:--:--
   --- ------------------------------------ 0.8/8.9 MB 2.1 MB/s eta 0:00:04
   ---- ----------------------------------- 1.0/8.9 MB 2.1 MB/s eta 0:00:04
   ------- -------------------------------- 1.6/8.9 MB 2.2 MB/s eta 0:00:04
   --------- ------------------------------ 2.1/8.9 MB 2.3 MB/s eta 0:00:04
   ---------- ----------------------------- 2.4/8.9 MB 2.3 MB/s eta 0:00:03
   -------------- ------------------------- 3.1/8.9 MB 2.3 MB/s eta 0:00:03
   --------------- ------------------------ 3.4/8.9 MB 2.2 MB/s eta 0:00:03
   ----------------- ---------------------- 3.9/8.9 MB 2.2 MB/s eta 0:00:03
   -------------------- ------------------- 4.5/8.9 MB 2.2 MB/s eta 0:00:03
   ---------------------- ----------------- 5.0/8.9 MB 2.2 MB/s eta 0:00:02
   ------------------------ -----


[notice] A new release of pip is available: 25.1.1 -> 25.2
[notice] To update, run: python.exe -m pip install --upgrade pip


In [3]:
df = pd.read_csv('Data\Churn_Modelling_New.csv')
df


RowNumber  CustomerId    Surname  CreditScore Geography  Gender   Age  \
0             1    15634602   Hargrave          619    France  Female  42.0   
1             2    15647311       Hill          608     Spain  Female  41.0   
2             3    15619304       Onio          502    France  Female  42.0   
3             4    15701354       Boni          699    France  Female  39.0   
4             5    15737888   Mitchell          850     Spain  Female  43.0   
...         ...         ...        ...          ...       ...     ...   ...   
9995       9996    15606229   Obijiaku          771    France    Male  39.0   
9996       9997    15569892  Johnstone          516    France    Male  35.0   
9997       9998    15584532        Liu          709    France  Female  36.0   
9998       9999    15682355  Sabbatini          772   Germany    Male  42.0   
9999      10000    15628319     Walker          792    France  Female  28.0   

      Tenure    Balance  NumOfProducts  HasCrCard  IsActiveMember  \
0          2       0.00              1          1               1   
1          1   83807.86              1          0               1   
2          8  159660.80              3          1               0   
3          1       0.00              2          0               0   
4          2  125510.82              1          1               1   
...      ...        ...            ...        ...             ...   
9995       5       0.00              2          1               0   
9996      10   57369.61              1          1               1   
9997       7       0.00              1          0               1   
9998       3   75075.31              2          1               0   
9999       4  130142.79              1          1               0   

      EstimatedSalary  Exited  
0           101348.88       1  
1           112542.58       0  
2           113931.57       1  
3            93826.63       0  
4            79084.10       0  
...               ...     ...  
9995         96270.64       0  
9996        101699.77       0  
9997         42085.58       1  
9998         92888.52       1  
9999         38190.78       0  

[10000 rows x 14 columns]

In [4]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 10000 entries, 0 to 9999
Data columns (total 14 columns):
 #   Column           Non-Null Count  Dtype  
---  ------           --------------  -----  
 0   RowNumber        10000 non-null  int64  
 1   CustomerId       10000 non-null  int64  
 2   Surname          10000 non-null  object 
 3   CreditScore      10000 non-null  int64  
 4   Geography        10000 non-null  object 
 5   Gender           9634 non-null   object 
 6   Age              9634 non-null   float64
 7   Tenure           10000 non-null  int64  
 8   Balance          10000 non-null  float64
 9   NumOfProducts    10000 non-null  int64  
 10  HasCrCard        10000 non-null  int64  
 11  IsActiveMember   10000 non-null  int64  
 12  EstimatedSalary  10000 non-null  float64
 13  Exited           10000 non-null  int64  
dtypes: float64(3), int64(8), object(3)
memory usage: 1.1+ MB


Missing values

In [6]:
df.isnull().sum()

RowNumber            0
CustomerId           0
Surname              0
CreditScore          0
Geography            0
Gender             366
Age                366
Tenure               0
Balance              0
NumOfProducts        0
HasCrCard            0
IsActiveMember       0
EstimatedSalary      0
Exited               0
dtype: int64

In [7]:

from sklearn.preprocessing import StandardScaler
from sklearn.preprocessing import MinMaxScaler

<b>What is Normalization? </b>

Normalization is a scaling technique in which values are shifted and rescaled so that they end up ranging between 0 and 1. It is also known as Min-Max scaling.



<b>What is Standardization?</b>

Standardization is another scaling technique where the values are centered around the mean with a unit standard deviation. This means that the mean of the attribute becomes zero and the resultant distribution has a unit standard deviation.

In [8]:
df.describe().round(2)

RowNumber   CustomerId  CreditScore      Age    Tenure    Balance  \
count   10000.00     10000.00     10000.00  9634.00  10000.00   10000.00   
mean     5000.50  15690940.57       650.53    38.95      5.01   76485.89   
std      2886.90     71936.19        96.65    10.49      2.89   62397.41   
min         1.00  15565701.00       350.00    18.00      0.00       0.00   
25%      2500.75  15628528.25       584.00    32.00      3.00       0.00   
50%      5000.50  15690738.00       652.00    37.00      5.00   97198.54   
75%      7500.25  15753233.75       718.00    44.00      7.00  127644.24   
max     10000.00  15815690.00       850.00    92.00     10.00  250898.09   

       NumOfProducts  HasCrCard  IsActiveMember  EstimatedSalary   Exited  
count       10000.00   10000.00        10000.00         10000.00  10000.0  
mean            1.53       0.71            0.52        100090.24      0.2  
std             0.58       0.46            0.50         57510.49      0.4  
min             1.00       0.00            0.00            11.58      0.0  
25%             1.00       0.00            0.00         51002.11      0.0  
50%             1.00       1.00            1.00        100193.92      0.0  
75%             2.00       1.00            1.00        149388.25      0.0  
max             4.00       1.00            1.00        199992.48      1.0

<b>Task-1 NORMALIZATION

In [9]:
df.head(5)

RowNumber  CustomerId   Surname  CreditScore Geography  Gender   Age  \
0          1    15634602  Hargrave          619    France  Female  42.0   
1          2    15647311      Hill          608     Spain  Female  41.0   
2          3    15619304      Onio          502    France  Female  42.0   
3          4    15701354      Boni          699    France  Female  39.0   
4          5    15737888  Mitchell          850     Spain  Female  43.0   

   Tenure    Balance  NumOfProducts  HasCrCard  IsActiveMember  \
0       2       0.00              1          1               1   
1       1   83807.86              1          0               1   
2       8  159660.80              3          1               0   
3       1       0.00              2          0               0   
4       2  125510.82              1          1               1   

   EstimatedSalary  Exited  
0        101348.88       1  
1        112542.58       0  
2        113931.57       1  
3         93826.63       0  
4         79084.10       0

In [ ]:
new_df = pd.DataFrame(df[['Age','Tenure']])
#new_df = pd.DataFrame(df,columns = ['Age', 'Tenure'])      Same as above
new_df.head(5)

Age  Tenure
0  42.0       2
1  41.0       1
2  42.0       8
3  39.0       1
4  43.0       2

In [12]:
new_df['Age'] = new_df['Age'].fillna(new_df['Age'].mean())

In [13]:
new_df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 10000 entries, 0 to 9999
Data columns (total 2 columns):
 #   Column  Non-Null Count  Dtype  
---  ------  --------------  -----  
 0   Age     10000 non-null  float64
 1   Tenure  10000 non-null  int64  
dtypes: float64(1), int64(1)
memory usage: 156.4 KB


<b>This the internal formula of MinMaxScaler
$$
x' = \frac{x - x_{\min}}{x_{\max} - x_{\min}}
$$


In [14]:
scaler = MinMaxScaler()
normalized_df = scaler.fit_transform(new_df)
print(normalized_df)

[[0.32432432 0.2       ]
 [0.31081081 0.1       ]
 [0.32432432 0.8       ]
 ...
 [0.24324324 0.7       ]
 [0.32432432 0.3       ]
 [0.13513514 0.4       ]]


In [15]:
x_array = np.array([[2],[3],[5],[6],[6]])

scaler = MinMaxScaler()
normalized_x = scaler.fit_transform(x_array)
print(normalized_x)


[[0.  ]
 [0.25]
 [0.75]
 [1.  ]
 [1.  ]]


<b>Task-2 STANDARDIZATION

$$
x′ = \frac{x - \mu}{\sigma}
$$

**Calculation:**

- Mean (μ) = 30
- Standard Deviation (σ) ≈ 15.81

$$
[ \frac{10-30}{15.81}, \frac{20-30}{15.81}, ..., \frac{50-30}{15.81} ] ≈ [-1.26, -0.63, 0.0, 0.63, 1.26]
$$

In [16]:
scaler = StandardScaler()
standardized_df = scaler.fit_transform(new_df)
print(standardized_df)


[[ 0.29588795 -1.04175968]
 [ 0.19878727 -1.38753759]
 [ 0.29588795  1.03290776]
 ...
 [-0.28671609  0.68712986]
 [ 0.29588795 -0.69598177]
 [-1.06352148 -0.35020386]]


In [18]:
x_array = np.array([[2],[3],[5],[6],[6]])

scaler = StandardScaler()
normalized_arr_ss = scaler.fit_transform(x_array)
print(normalized_arr_ss)

[[-1.47709789]
 [-0.86164044]
 [ 0.36927447]
 [ 0.98473193]
 [ 0.98473193]]


--------
Z-score Standardization Formula
$$
x′ = \frac{x - \mu}{\sigma}
$$
Where:  \
μ is the mean of the data   \
σ is the standard deviation


-------
1. Calculate the mean (μ)
$$
μ = \frac{2 + 3 + 5 + 6 + 6}{5} = \frac{22}{5} = 4.4
$$
-------
2. Calculate the standard deviation (σ)

Use **population standard deviation** (ddof = 0):

$$
σ= \sqrt{ \frac{(2-4.4)^2 + (3-4.4)^2 + (5-4.4)^2 + (6-4.4)^2 + (6-4.4)^2}{5} }
$$

$$
σ = \sqrt{ \frac{13.2}{5} } = \sqrt{2.64} \approx 1.6248
$$
---------
3. apply This Formula to each value
$$
x′ = \frac{x - 4.4}{1.6248}
$$


| x | x' = (x - 4.4) / 1.6248 | Rounded x' |
| --- | --- | --- |
| 2 | (2 - 4.4)/1.6248   |-1.477  |
| 3 | (3 - 4.4)/1.6248   |-0.862  |
| 5 | (5 - 4.4)/1.6248   |0.369  |
| 6 | (6 - 4.4)/1.6248   |0.985  |
| 6 | (6 - 4.4)/1.6248   |0.985  |


In [20]:
print(normalized_arr_ss)

[[-1.47709789]
 [-0.86164044]
 [ 0.36927447]
 [ 0.98473193]
 [ 0.98473193]]
